File: typescript/src/main/vsdk/toolkit/raycasting/common/SimpleRaytracingPathNode.ts
import { PrintStream } from "../../../../java/io/PrintStream";
import { ColorRgb } from "../../common/ColorRgb";
import { Error as VsdkError } from "../../common/Error";
import { Vector3D } from "../../common/linealAlgebra/Vector3D";
import { Vector3DPrinter } from "../../io/wrapper/Vector3DPrinter";
import { PhongBidirectionalScatteringDistributionFunction } from "../../material/PhongBidirectionalScatteringDistributionFunction";
import { RayHitFlag } from "../../material/RayHitFlag";
import { RayHit } from "../../skin/RayHit";
import { BsdfComp } from "./BsdfComp";
import { PathRayType } from "./PathRayType";

export class SimpleRaytracingPathNode {
  public static multipleImportanceSampling(a: number): number {
    return a * a;
  }

  public m_hit: RayHit;
  public m_inDirT: Vector3D;
  public m_inDirF: Vector3D;
  public m_normal: Vector3D;

  public m_G: number;
  public m_pdfFromPrev: number;
  public m_pdfFromNext: number;
  public m_rrPdfFromNext: number;
  public accumulatedRussianRouletteFactors: number;

  public m_bsdfEval: ColorRgb;
  public m_bsdfComp: BsdfComp;

  public m_usedComponents: number;
  public m_accUsedComponents: number;

  public m_useBsdf: PhongBidirectionalScatteringDistributionFunction | null;
  public m_inBsdf: PhongBidirectionalScatteringDistributionFunction | null;
  public m_outBsdf: PhongBidirectionalScatteringDistributionFunction | null;
  public m_rayType: PathRayType;
  public m_depth: number;

  private m_next: SimpleRaytracingPathNode | null;
  private m_previous: SimpleRaytracingPathNode | null;

  public constructor() {
    this.m_hit = new RayHit();
    this.m_inDirT = new Vector3D();
    this.m_inDirF = new Vector3D();
    this.m_normal = new Vector3D();

    this.m_G = 0.0;
    this.m_pdfFromPrev = 0.0;
    this.m_pdfFromNext = 0.0;
    this.m_rrPdfFromNext = 0.0;
    this.accumulatedRussianRouletteFactors = 0.0;

    this.m_bsdfEval = new ColorRgb();
    this.m_bsdfComp = new BsdfComp();

    this.m_usedComponents = 0;
    this.m_accUsedComponents = 0;
    this.m_useBsdf = null;
    this.m_inBsdf = null;
    this.m_outBsdf = null;
    this.m_rayType = PathRayType.STARTS;
    this.m_depth = 0;

    this.m_next = null;
    this.m_previous = null;
  }

  public next(): SimpleRaytracingPathNode | null {
    return this.m_next;
  }

  public previous(): SimpleRaytracingPathNode | null {
    return this.m_previous;
  }

  public setNext(node: SimpleRaytracingPathNode | null): void {
    this.m_next = node;
  }

  public setPrevious(node: SimpleRaytracingPathNode | null): void {
    this.m_previous = node;
  }

  public attach(node: SimpleRaytracingPathNode): void {
    this.m_next = node;
    node.setPrevious(this);
  }

  public ensureNext(): void {
    if (this.m_next === null) {
      this.attach(new SimpleRaytracingPathNode());
    }
  }

  public print(out: PrintStream | null): void {
    if (out === null) {
      return;
    }

    out.printf("Path node at depth %d\n", this.m_depth);
    out.printf("Pos : ");
    Vector3DPrinter.print(this.m_hit.getPoint(), out);
    out.printf("\n");
    out.printf("Norm: ");
    Vector3DPrinter.print(this.m_normal, out);
    out.printf("\n");
    if (this.m_previous !== null) {
      out.printf("InF: ");
      Vector3DPrinter.print(this.m_inDirF, out);
      out.printf("\n");
      out.printf("Cos in  %f\n", this.m_normal.dotProduct(this.m_inDirF));
      const patch = this.m_hit.getPatch();
      if (patch !== null) {
        out.printf("GCos in %f\n", patch.normal.dotProduct(this.m_inDirF));
      }
    }
    if (this.m_next !== null) {
      out.printf("OutF: ");
      Vector3DPrinter.print(this.m_next.m_inDirT, out);
      out.printf("\n");
      out.printf("Cos out %f\n", this.m_normal.dotProduct(this.m_next.m_inDirT));
      const patch = this.m_hit.getPatch();
      if (patch !== null) {
        out.printf("GCos out %f\n", patch.normal.dotProduct(this.m_next.m_inDirT));
      }
    }
  }

  protected GetMatchingNode(): SimpleRaytracingPathNode | null {
    const thisBsdf = this.m_useBsdf;
    let backHits = 1;
    let tmpNode = this.previous();
    let matchedNode: SimpleRaytracingPathNode | null = null;

    while (tmpNode !== null && backHits > 0) {
      switch (tmpNode.m_rayType) {
        case PathRayType.ENTERS: {
          const hitPatch = tmpNode.m_hit.getPatch();
          const hitBsdf = hitPatch !== null && hitPatch.material !== null
            ? hitPatch.material.getBsdf()
            : null;
          if (hitBsdf === thisBsdf) {
            backHits--;
          }
          break;
        }
        case PathRayType.LEAVES:
          if (tmpNode.m_inBsdf === thisBsdf) {
            backHits++;
          }
          break;
        case PathRayType.REFLECTS:
          break;
        default:
          VsdkError.error("CPathNode::GetMatchingNode", "Wrong ray type in path");
      }

      matchedNode = tmpNode;
      tmpNode = tmpNode.previous();
    }

    return backHits === 0 ? matchedNode : null;
  }

  public getPreviousBsdf(): PhongBidirectionalScatteringDistributionFunction | null {
    if ((this.m_hit.getFlags() & RayHitFlag.BACK) === 0) {
      VsdkError.error("CPathNode::getPreviousBsdf", "Last node not a back hit");
      return this.m_inBsdf;
    }

    const patch = this.m_hit.getPatch();
    const patchBsdf = patch !== null && patch.material !== null ? patch.material.getBsdf() : null;
    if (patchBsdf !== this.m_inBsdf) {
      VsdkError.warning("CPathNode::GetPreviousBtdf", "Last back hit has wrong bsdf");
    }

    const matchedNode = this.GetMatchingNode();

    if (matchedNode === null) {
      VsdkError.warning("CPathNode::GetPreviousBtdf", "No corresponding entering ray");
      return this.m_inBsdf;
    }

    return matchedNode.m_inBsdf;
  }

  public assignBsdfAndNormal(): void {
    const patch = this.m_hit.getPatch();
    if (patch === null || patch.material === null) {
      return;
    }

    this.m_normal.copy(this.m_hit.getNormal());

    this.m_useBsdf = patch.material.getBsdf();

    if ((this.m_hit.getFlags() & RayHitFlag.FRONT) !== 0) {
      this.m_outBsdf = this.m_useBsdf;
    }
    else {
      this.m_outBsdf = this.getPreviousBsdf();
    }
  }

  public ends(): boolean {
    return this.m_rayType === PathRayType.STOPS || this.m_rayType === PathRayType.ENVIRONMENT;
  }
}

File: java/src/main/java/vsdk/toolkit/raycasting/common/SimpleRaytracingPathNode.java
/**
Class definition of path nodes. These node are building blocks of paths.
and contain necessary information for raytracing-like algorithms
*/

package vsdk.toolkit.raycasting.common;

import java.io.PrintStream;

import vsdk.toolkit.common.logging.Logger;
import vsdk.toolkit.common.linealAlgebra.Vector3D;
import vsdk.toolkit.common.color.ColorRgb;
import vsdk.toolkit.io.wrapper.Vector3DPrinter;
import vsdk.toolkit.material.Material;
import vsdk.toolkit.material.PhongBidirectionalScatteringDistributionFunction;
import vsdk.toolkit.skin.RayHitFlag;
import vsdk.toolkit.skin.RayHit;

// Type definitions used in CPathNode

// -- TODO clean up, additional functions that are now duplicated
// -- in the samplers, accessor methods, splitting in a
// -- path node and spear rate connections

public class SimpleRaytracingPathNode {
    /**
    Heuristic for multiple Importance sampling / weighting
    */
    public static double multipleImportanceSampling(double a) {
        return a * a;
    }

    public RayHit m_hit;
    public Vector3D m_inDirT; // towards the patch
    public Vector3D m_inDirF; // from the patch
    public Vector3D m_normal;

    public double m_G; // Geometric factor x(i-1) -> x(i)
    public double m_pdfFromPrev;
    public double m_pdfFromNext;
    public double m_rrPdfFromNext; // Path length in other direction not
    //  known beforehand => separate components needed.
    public double accumulatedRussianRouletteFactors;

    public ColorRgb m_bsdfEval;
    public BsdfComp m_bsdfComp;

    public byte m_usedComponents; // Components used for scattering in this
    // node. For full bsdf sampling, these are all components, independent
    // of which component is actually used (M.I.S.), but other types
    // of sampling are possible.
    public byte m_accUsedComponents; // Accumulated used components. This allows
    // one to check immediately if any previous node in the path had a certain
    // scattering component used. accUsed = prev->used | prev->accUsed

    public PhongBidirectionalScatteringDistributionFunction m_useBsdf; // bsdf used for scattering
    public PhongBidirectionalScatteringDistributionFunction m_inBsdf; // Medium of incoming ray
    public PhongBidirectionalScatteringDistributionFunction m_outBsdf; //Medium of a possible transmitted ray (other side of normal)
    public PathRayType m_rayType;
    public int m_depth; // First node in a path has depth 0

    private SimpleRaytracingPathNode m_next;
    private SimpleRaytracingPathNode m_previous;

    public SimpleRaytracingPathNode() {
        m_hit = new RayHit();
        m_inDirT = new Vector3D();
        m_inDirF = new Vector3D();
        m_normal = new Vector3D();

        m_G = 0.0;
        m_pdfFromPrev = 0.0;
        m_pdfFromNext = 0.0;
        m_rrPdfFromNext = 0.0;
        accumulatedRussianRouletteFactors = 0.0;

        m_bsdfEval = new ColorRgb();
        m_bsdfComp = new BsdfComp();

        m_usedComponents = 0;
        m_accUsedComponents = 0;
        m_useBsdf = null;
        m_inBsdf = null;
        m_outBsdf = null;
        m_rayType = PathRayType.STARTS;
        m_depth = 0;

        m_next = null;
        m_previous = null;
    }

    // Navigation in a path
    public SimpleRaytracingPathNode next() {
        return m_next;
    }

    public SimpleRaytracingPathNode previous() {
        return m_previous;
    }

    public void setNext(SimpleRaytracingPathNode node) {
        m_next = node;
    }

    public void setPrevious(SimpleRaytracingPathNode node) {
        m_previous = node;
    }

    public void attach(SimpleRaytracingPathNode node) {
        m_next = node;
        node.setPrevious(this);
    }

    public void ensureNext() {
        if (m_next == null) {
            attach(new SimpleRaytracingPathNode());
        }
    }

    public void print(PrintStream out) {
        if (out == null) {
            return;
        }

        out.printf("Path node at depth %d\n", m_depth);
        out.printf("Pos : ");
        Vector3DPrinter.print(m_hit.getPoint(), out);
        out.printf("\n");
        out.printf("Norm: ");
        Vector3DPrinter.print(m_normal, out);
        out.printf("\n");
        if (m_previous != null) {
            out.printf("InF: ");
            Vector3DPrinter.print(m_inDirF, out);
            out.printf("\n");
            out.printf("Cos in  %f\n", m_normal.dotProduct(m_inDirF));
            out.printf("GCos in %f\n", m_hit.getPatch().normal.dotProduct(m_inDirF));
        }
        if (m_next != null) {
            out.printf("OutF: ");
            Vector3DPrinter.print(m_next.m_inDirT, out);
            out.printf("\n");
            out.printf("Cos out %f\n", m_normal.dotProduct(m_next.m_inDirT));
            out.printf("GCos out %f\n", m_hit.getPatch().normal.dotProduct(m_next.m_inDirT));
        }
    }

    /**
GetPreviousBsdf

Searches back in the path to find the bsdf on the outside
of the current path/material. The last node should be
a back hit -> Leaving ray-type. If not, the current bsdf
is returned and an error is reported
*/

    /**
Helper routine, searches the corresponding 'Enters' node for this node
*/
    protected SimpleRaytracingPathNode GetMatchingNode() {
        final PhongBidirectionalScatteringDistributionFunction thisBsdf = m_useBsdf;
        int backHits = 1;
        SimpleRaytracingPathNode tmpNode = previous();
        SimpleRaytracingPathNode matchedNode = null;

        while (tmpNode != null && backHits > 0) {
            switch (tmpNode.m_rayType) {
                case ENTERS:
                    if (tmpNode.m_hit.getPatch().material.getBsdf() == thisBsdf) {
                        backHits--; // Entering point in this material
                    }
                    break;
                case LEAVES:
                    if (tmpNode.m_inBsdf == thisBsdf) {
                        backHits++; // Leaves the same material more than one time
                    }
                    break;
                case REFLECTS:
                    break;
                default:
                    Logger.error("CPathNode::GetMatchingNode", "Wrong ray type in path");
            }

            matchedNode = tmpNode;
            tmpNode = tmpNode.previous();
        }

        if (backHits == 0) {
            return matchedNode;
        }
        else {
            return null;  // No matching node
        }
    }

    public PhongBidirectionalScatteringDistributionFunction getPreviousBsdf() {
        if ((m_hit.getFlags() & RayHitFlag.BACK) == 0) {
            Logger.error("CPathNode::getPreviousBsdf", "Last node not a back hit");
            return m_inBsdf;  // Should not happen
        }

        if (m_hit.getPatch().material.getBsdf() != m_inBsdf) {
            Logger.warning("CPathNode::GetPreviousBtdf", "Last back hit has wrong bsdf");
        }

        // Find the corresponding ray that enters the material
        final SimpleRaytracingPathNode matchedNode = GetMatchingNode();

        if (matchedNode == null) {
            Logger.warning("CPathNode::GetPreviousBtdf", "No corresponding entering ray");
            return m_inBsdf;  // Should not happen
        }

        return matchedNode.m_inBsdf;
    }

    public void assignBsdfAndNormal() {
        final Material thisMaterial;

        if (m_hit.getPatch() == null) {
            // Invalid node
            return;
        }

        thisMaterial = m_hit.getPatch().material;

        m_normal.copy(m_hit.getNormal()); // Possible double format

        // Assign bsdf's
        m_useBsdf = thisMaterial.getBsdf();

        if ((m_hit.getFlags() & RayHitFlag.FRONT) != 0) {
            m_outBsdf = m_useBsdf; // In filled in when creating this node
        }
        else {
            // BACK HIT
            m_outBsdf = getPreviousBsdf();
        }
    }

    public boolean ends() {
        return (m_rayType == PathRayType.STOPS) || (m_rayType == PathRayType.ENVIRONMENT);
    }
}
